rescript/ncbi.py

```diff
@@ -88,87 +114,216 @@ def get_ncbi_data(
+            request_lock.release()
+            logger.debug('request lock released')
+        r.raise_for_status()
+        data = parse(r.content)
+        data = list(data.values()).pop()
+        data = list(data.values()).pop()
+        # check that we got everything
+        if isinstance(data, list):
+            for rec in data:
+                if not isinstance(rec, dict):
+                    logger.debug('bad record:\n' + str(rec))
+                    raise RuntimeError('bad record')
+            else:
+                return data
+        elif isinstance(data, dict):
+            return [data]
+        else:
+            logger.debug('bad record:\n' + str(data))
+            raise RuntimeError('bad record')
+    return _robustify(request, logger)
+
+
+def _large_warning(logging_level):
+    _get_logger(logging_level).warning(
+        'This query could result in more than 100 requests to NCBI. If you '
+        'are not running it on the weekend or between 9 pm and 5 am Eastern '
+        'Time weekdays, it may result in NCBI blocking your IP address. See '
+        'https://www.ncbi.nlm.nih.gov/home/about/policies/ for details.')
+
+
+def _ungotten_ids(ids, data):
+    error = ("Partial download. Expected " + str(len(ids)) +
+             ' records, but got ' + str(len(data)) + '.')
+    id_keys = ['TSeq_accver', 'TaxId']
+    gotten = set()
+    for record in data:
+        for id_key in id_keys:
+            if id_key in record:
+                gotten.add(record[id_key])
+                break
+    ungotten = list(set(ids) - gotten)
+    if len(ungotten) > 10:
+        error += ('\nMore than 10 ids were missing. Ten were: ')
+    else:
+        error += '\nThe following ids were missing: '
+    error += ', '.join(ungotten[:10]) + '.\n'
+    return error
+
+
+def _get_id_chunk(
+        ids_chunk, params, request_lock, logging_level, raise_on_partial,
+        entrez_delay):
+    chunk_params = _epost(
+        params, ids_chunk, request_lock, logging_level, entrez_delay)
+    data_chunk = _efetch_5000(
+        chunk_params, request_lock, logging_level, entrez_delay)
+    logger = _get_logger(logging_level)
+    if len(ids_chunk) != len(data_chunk):
+        error = _ungotten_ids(ids_chunk, data_chunk)
+        if raise_on_partial:
             raise RuntimeError(error)
-        chunk = parse(r.content)
-        chunk = list(chunk.values()).pop()
-        chunk = list(chunk.values()).pop()
-        if isinstance(chunk, list):
-            data.extend(chunk)
         else:
-            data.append(chunk)
+            logger.warning(error)
+    logger.info('got ' + str(len(data_chunk)) + ' records')
+    return data_chunk
+
+
+def _get_for_ids(params, ids, logging_level, n_jobs, request_lock,
+                 raise_on_partial, entrez_delay=0.334):
+    logger = _get_logger(logging_level)
+    logger.info('Downloading ' + str(len(ids)) + ' records')
+    ids = list(ids)
+    parallel = Parallel(n_jobs=n_jobs, backend='loky')
```

Comment:
My logic went like this: joblib will be used elsewhere in this library so I will use joblib (I also like it). Now I need a process lock and the one that has the best chance of working is the one that is implemented in joblib, which resides in the loky backend.

The extra overhead incurred by using multiprocessing rather than multithreading isn't the bottleneck in this program. NCBI is the bottleneck.

Reply:
Oh, my point wasn't against joblib. I only meant that the `Parallel` manager uses multprocessing when backend is set to "loky" but can also use multithreading if set to something else ("multithreading"? 😅 ). But I guess you're right, either one should be fine.

Reply:
Thanks @misialq, I meant that I chose loky and loky locks because I they look like they've been developed with joblib, whereas I was less confident about any of the other backends.